### PR TITLE
UCP/SOCKADDR/CM: fix ucp_cm_server_conn_notify_cb err handling

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
## What
do not reset `UCP_EP_FLAG_LOCAL_CONNECTED` flag on err handling flow of `ucp_cm_server_conn_notify_cb` to invoke `uct_ep_disconnect` from `ucp_worker_set_ep_failed` to invoke `disconnect_cb` on client side and complete close request.

## Why ?
fixes https://github.com/openucx/ucx/issues/5162

### CODE REVIEW

File: src/ucp/wireup/wireup_cm.c [hunk 1]
```diff
@@ -520,7 +520,8 @@ static unsigned ucp_ep_cm_disconnect_progress(void *arg)
 
     UCS_ASYNC_BLOCK(async);
 
-    ucs_trace("ep %p: got remote disconnect, cm_ep %p", ucp_ep, uct_cm_ep);
+    ucs_trace("ep %p: got remote disconnect, cm_ep %p, flags 0x%x", ucp_ep,
+             uct_cm_ep, ucp_ep->flags);
```
Comment (hoopoepg) [inline src/ucp/wireup/wireup_cm.c:524]: alignment

Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c]: thanks, fixed

File: src/ucp/wireup/wireup_cm.c [hunk 2]
```diff
@@ -855,7 +856,6 @@ static void ucp_cm_server_conn_notify_cb(uct_ep_h ep, void *arg,
         /* if reject is arrived on server side, then UCT does something wrong */
         ucs_assert(status != UCS_ERR_REJECTED);
         cm_lane = ucp_ep_get_cm_lane(ucp_ep);
-        ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:858]: what was the reason it was added in the first place?

**Comment by yosefe** [comment]
@evgeny-leksikov  can failure be related?
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/6092/logs/111
```
2020-05-14T17:02:57.9218662Z [     INFO ] server listening on fe80::268a:7ff:fe8d:af27:45147
2020-05-14T17:02:58.0420527Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:283: Failure
2020-05-14T17:02:58.0422610Z Failed
2020-05-14T17:02:58.0423335Z Got 1 warnings during the test
2020-05-14T17:02:58.0424176Z [  FAILED  ] dc_ud/test_ucp_sockaddr_destroy_ep_on_err.onesided_bidi_sforce/7, where GetParam() = dc_x,ud_v,ud_x,mm (413 ms)
```

**Comment by evgeny-leksikov** [comment]
> @evgeny-leksikov can failure be related?
> https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/6092/logs/111
> 
> ```
> 2020-05-14T17:02:57.9218662Z [     INFO ] server listening on fe80::268a:7ff:fe8d:af27:45147
> 2020-05-14T17:02:58.0420527Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:283: Failure
> 2020-05-14T17:02:58.0422610Z Failed
> 2020-05-14T17:02:58.0423335Z Got 1 warnings during the test
> 2020-05-14T17:02:58.0424176Z [  FAILED  ] dc_ud/test_ucp_sockaddr_destroy_ep_on_err.onesided_bidi_sforce/7, where GetParam() = dc_x,ud_v,ud_x,mm (413 ms)
> ```

looking at this (cant reproduce yet) but unlikely because this test establishes connection and does bidirectional communication, then testing close protocol with force close. this PR affects only err handling on wireup

File: src/ucp/wireup/wireup_cm.c (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:858]: most likely copy-past programming.. reset of the flag before set_ep_failed makes sense for transport failure and after getting disconnect callback to avoid "double" disconnect

Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:858]: i see other places which drop LOCAL_CONNECTED flag, should they be updated as well?

Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:858]: actually it's hard to answer definitely and cover all possible cases without regular testing of link failure because one-sided disconnect with live link and transport error without possibility to get response have different protocols flow and have to be handled by user different way. IMO it's better to investigate cases one-by-one.

Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:858]: i'm confused, because https://github.com/openucx/ucx/pull/5075/files#diff-36e34c1df2a6c773b9dda08dd8bf17eeR328 just added the dropping of LOCAL_CONNECTED flag before set_ep_failed, so which one it should be?
IMO we should never drop because ucp_worker_set_ep_failed() calls ucp_ep_cm_disconnect_cm_lane() if needed

File: src/uct/ib/rdmacm/rdmacm_cm.c [hunk 3]
```diff
@@ -300,8 +300,6 @@ static void uct_rdmacm_cm_handle_event_connect_response(struct rdma_cm_event *ev
                   ucs_sockaddr_str(remote_addr, ip_port_str,
                                    UCS_SOCKADDR_STRING_LEN));
         uct_rdmacm_cm_ep_set_failed(cep, &remote_data, status);
-        /* notify remote side about local error */
-        rdma_disconnect(cep->id);
```
Comment (alinask) [inline src/uct/ib/rdmacm/rdmacm_cm.c:304]: @evgeny-leksikov it's removed to prevent a second rdma_disconnect call since uct_rdmacm_cm_ep_set_failed() will invoke it?

Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm.c:304]: it's user (UCP) responsibility how to handle error - disconnect or just destroy UCT ep (including rdmacm id) if wireup is not completed

**Comment by evgeny-leksikov** [comment]
bot:pipe:retest
setup issue

File: src/ucp/wireup/wireup_cm.c [hunk 4]
```diff
@@ -312,24 +311,23 @@ static unsigned ucp_cm_client_connect_progress(void *arg)
 
     status = uct_cm_client_ep_conn_notify(uct_cm_ep);
     if (status != UCS_OK) {
+        ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:314]: isn't set_ep_failed() doing this already?
seems removing this flag will prevent set_ep_failed from removing CM lane, why is this what we want?

File: src/ucp/wireup/wireup_cm.c [hunk 5]
```diff
@@ -417,9 +416,10 @@ static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
 err_free_arg:
     ucs_free(progress_arg);
 err_out:
-    ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
+    UCS_ASYNC_BLOCK(&worker->async);
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:418]: maybe we need to put lock inside ucp_worker_set_ep_failed?
or at least comment it should be called with lock held?

File: src/ucp/wireup/wireup_cm.c [hunk 6]
```diff
@@ -370,6 +368,7 @@ static void ucp_cm_client_connect_cb(uct_ep_h uct_cm_ep, void *arg,
     status      = connect_args->status;
 
     if (status != UCS_OK) {
+        ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
```
Comment (yosefe) [inline src/ucp/wireup/wireup_cm.c:370]: same question - why in some cases we remove UCP_EP_FLAG_LOCAL_CONNECTED before calling set_ep_failed, and in others don't?

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c [hunk 7]
```diff
@@ -110,6 +110,13 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
               uct_rdmacm_cm_ep_str(cep, ep_str, UCT_RDMACM_EP_STRING_LEN),
               cep->id, rdmacm_cm, rdmacm_cm->ev_ch);
 
+    if (cep->flags & (UCT_RDMACM_CM_EP_FAILED |
+                      UCT_RDMACM_CM_EP_GOT_DISCONNECT)) {
+        remote_data.field_mask = 0;
+        uct_rdmacm_cm_ep_set_failed(cep, &remote_data, cep->status);
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm_ep.c:116]: why is this needed? we should have already called disconnect callback when set the FAILED flag?
should just return error in this case

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c [hunk 8]
```diff
@@ -53,10 +53,6 @@ void uct_rdmacm_cm_ep_error_cb(uct_rdmacm_cm_ep_t *cep,
                                uct_cm_remote_data_t *remote_data,
                                ucs_status_t status)
 {
-    if (cep->flags & UCT_RDMACM_CM_EP_FAILED) {
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm_ep.c:56]: so now we don't have check that uct_rdmacm_cm_handle_event_disconnected() is not calling disconnect callback more than once?

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c [hunk 9]
```diff
@@ -372,12 +379,18 @@ ucs_status_t uct_rdmacm_cm_ep_disconnect(uct_ep_h ep, unsigned flags)
...
-        ucs_error("%s: id=%p to peer %s",
-                  uct_rdmacm_cm_ep_str(cep, ep_str, UCT_RDMACM_EP_STRING_LEN),
-                  cep->id, ucs_sockaddr_str(rdma_get_peer_addr(cep->id),
-                                            ip_port_str,
-                                            UCS_SOCKADDR_STRING_LEN));
-        status = cep->status;
+        if (cep->status == UCS_ERR_CONNECTION_RESET) {
+            /* the ep was failed during wireup, remote side is not connected
+             * or destroyed */
+            status = UCS_OK;
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm_ep.c:385]: why UCS_OK? caller should be issue disconnect on a failed endpoint

File: src/ucp/wireup/wireup_cm.c (hunk 4 above)
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:314]: if the flag is set set_ep_failed() will call disconnect on failed EP, this is not what we want (UCT will return error)
(failed and disconnected lane is not the same)

File: src/ucp/wireup/wireup_cm.c (hunk 6 above)
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:370]: we remove this flag is wireup is not completed - connection establishment failed.
if one from EPs has completed connection - other side has to call disconect.

File: src/ucp/wireup/wireup_cm.c (hunk 5 above)
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:418]: it's commented, in many cases we already have acquired lock

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 8 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:56]: the check is moved to the same place where the flag is set - uct_rdmacm_cm_ep_set_failed

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 7 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:116]: connect notify and error callback can be invoked from different threads, so this is needed to report error by err callback immediately

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 9 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:385]: if the EP failed during wireup then should not

File: src/ucp/wireup/wireup_cm.c (hunk 5 above)
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:418]: where is it commented?

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 8 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:56]: but uct_rdmacm_cm_handle_event_disconnected() also calls uct_rdmacm_cm_ep_error_cb()

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 7 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:116]: what is the purpose of calling uct_rdmacm_cm_ep_set_failed() when FAILED flag is set, if uct_rdmacm_cm_ep_set_failed() not doing anything if FAILED flag is set (line 78)?

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 9 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:385]: but in UCP, you remove LOCAL_CONNECTED flag, in order to avoid calling disconnect on a failed endpoint per your comment https://github.com/openucx/ucx/pull/5166/files#r427789163). So why need to avoid error report on UCT side?

File: src/ucp/core/ucp_worker.c [hunk 10]
```diff
@@ -503,6 +503,9 @@ int ucp_worker_err_handle_remove_filter(const ucs_callbackq_elem_t *elem,
     return 0;
 }
 
+/*
+ * Caller must aquire lock
```
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c:507]: typo: aquire -> acquire

File: src/ucp/wireup/wireup_cm.c (hunk 5 above)
Comment (evgeny-leksikov) [reply src/ucp/wireup/wireup_cm.c:418]: https://github.com/openucx/ucx/pull/5166/files#r427865158

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 8 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:56]: disconnected != failed, if the EP is got disconnect event but valid, user have to call disconnect as well, if the EP failed during wiruep, then no need to notify remote side.

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 7 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:116]: to avoid code duplication since we have protection for failed state and move the EP to failed state on disconnect event during wireup

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c (hunk 9 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm_ep.c:385]: this case is caused by uct_flush specific, it's possible when UCP do not send anything on TL lanes flushed empty lanes and interrupts wireup by destroying all lanes including CM lane

**Comment by evgeny-leksikov** [comment]
setup failure
bot:pipe:retest

**Comment by evgeny-leksikov** [comment]
bot:pipe:retest

File: src/ucp/wireup/wireup_cm.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/wireup/wireup_cm.c:314]: pls add code comment to explain

File: src/uct/ib/rdmacm/rdmacm_cm.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/rdmacm/rdmacm_cm.c:304]: i guess this is needed because of https://github.com/openucx/ucx/pull/5166/files#diff-36e34c1df2a6c773b9dda08dd8bf17eeL858?

File: src/uct/ib/rdmacm/rdmacm_cm_ep.c [hunk 11]
```diff
@@ -110,6 +110,15 @@ ucs_status_t uct_rdmacm_cm_ep_conn_notify(uct_ep_h ep)
...
               cep->id, rdmacm_cm, rdmacm_cm->ev_ch);
 
+    UCS_ASYNC_BLOCK(uct_rdmacm_cm_ep_get_async(cep));
+    if (cep->flags & (UCT_RDMACM_CM_EP_FAILED |
+                      UCT_RDMACM_CM_EP_GOT_DISCONNECT)) {
+        UCS_ASYNC_UNBLOCK(uct_rdmacm_cm_ep_get_async(cep));
+        return cep->status;
+    }
+
+    UCS_ASYNC_UNBLOCK(uct_rdmacm_cm_ep_get_async(cep));
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm_ep.c:120]: looks like uct_rdmacm_cm_ep_set_failed() can be called outside the lock?

File: src/uct/ib/rdmacm/rdmacm_cm.c (hunk 3 above)
Comment (evgeny-leksikov) [reply src/uct/ib/rdmacm/rdmacm_cm.c:304]: yes, now it's unified and controlled by upper layer (UCP)

**Comment by yosefe** [comment]
bot:pipe:retest
